### PR TITLE
MCOL-641 1. Minor refactoring of decimalToString for int128_t.

### PR DESCRIPTION
**Pull request by tntnatbry**
2. Update unit tests for decimalToString.
3. Allow support for wide decimal in TupleConstantStep::fillInConstants().

### CODE REVIEW

**Comment by tntnatbry** [comment]
With this change, the functional tests for inserts are now passing.

File: dbcon/joblist/jlf_execplantojoblist.cpp [hunk 1]
```diff
@@ -1880,7 +1880,9 @@ const JobStepVector doSimpleFilter(SimpleFilter* sf, JobInfo& jobInfo)
             if (ct.colDataType == CalpontSystemCatalog::DECIMAL &&
                 ct.colWidth == 16)
             {
-                dataconvert::atoi128(constval, val128);
+                bool saturate = false;
+                val128 = dataconvert::string_to_ll<int128_t>(constval, saturate);
+                // TODO MCOL-641 check saturate
```
Comment (drrtuy) [inline dbcon/joblist/jlf_execplantojoblist.cpp]: MDB doesn't neither return the warning nor do anything if the value is saturated. It doesn't allow saturable value in a SET part also.

File: dbcon/joblist/pcolscan.cpp [hunk 2]
```diff
@@ -176,7 +176,8 @@ pColScanStep::pColScanStep(
     // MCOL-641 WIP
     else if (fColType.colWidth > 8 
         && fColType.colDataType != CalpontSystemCatalog::BINARY
-        && fColType.colDataType != CalpontSystemCatalog::DECIMAL)
+        && fColType.colDataType != CalpontSystemCatalog::DECIMAL
+        && fColType.colDataType != CalpontSystemCatalog::UDECIMAL)
```
Comment (drrtuy) [inline dbcon/joblist/pcolscan.cpp]: So these were the errors that broke UNSIGNED.

File: tests/dataconvert-tests.cpp [hunk 3]
```diff
@@ -545,126 +545,189 @@ TEST(DataConvertTest, NumberIntValue)
...
-    uint8_t scale3 = 3;
-    decimalAsStringVec.push_back("99999999999999999999999999999999999999");
-    decimalAsStringVec.push_back("-99999999999999999999999999999999999999");
-    decimalAsStringVec.push_back("0");
-    decimalAsStringVec.push_back("-0.042");
-    expected.push_back("99999999999999999999999999999999999.999");
-    expected.push_back("-99999999999999999999999999999999999.999");
-    expected.push_back("0.000");
-    expected.push_back("-0.042");
+TEST(DataConvertTest, DecimalToStringCheckScale0)
```
Comment (drrtuy) [inline tests/dataconvert-tests.cpp]: Could you ellaborate on your reasoning behind the change?

File: utils/dataconvert/dataconvert.cpp [hunk 4]
```diff
@@ -1249,115 +1250,144 @@ bool stringToTimestampStruct(const string& data, TimeStamp& timeStamp, const str
...
+
+    if (scale)
+    {
+        const uint8_t maxPowOf10 =
+            (sizeof(columnstore_pow_10) / sizeof(columnstore_pow_10[0])) - 1;
+
+        // Assuming scale = [0, 56]
+        switch (scale / maxPowOf10)
+        {
+            case 2: // scale = [38, 56]
```
Comment (drrtuy) [inline utils/dataconvert/dataconvert.cpp]: I wonder why do you change case to unparenthised here and later use the other way?

File: utils/dataconvert/dataconvert.cpp [hunk 5]
```diff
@@ -1379,54 +1409,20 @@ void DataConvert::toString(int128_t* dec, uint8_t scale,
     if (scale)
     {
         *p++ = '.';
-        p += writeFractionalPart(dec, p, buflen-(p-original_p), scale);
+        written = p - original_p;
```
Comment (drrtuy) [inline utils/dataconvert/dataconvert.cpp]: Just increment written.

**Comment by drrtuy** [comment]
The patch is based on one commit before the current HEAD so plz rebase it.

File: utils/rowgroup/rowgroup.h [hunk 6]
```diff
@@ -433,6 +433,9 @@ class Row
     // that's not string-table safe, this one is
     inline void copyField(Row& dest, uint32_t destIndex, uint32_t srcIndex) const;
 
+    // WIP MCOL-641
+    inline void copyBinaryField(Row& dest, uint32_t destIndex, uint32_t srcIndex) const;
```
Comment (drrtuy) [inline utils/rowgroup/rowgroup.h]: Kindly add the corresponding UT.

File: dbcon/joblist/jlf_execplantojoblist.cpp (hunk 1 above)
Comment (tntnatbry) [reply dbcon/joblist/jlf_execplantojoblist.cpp]: This is GTK. Is this specific to a query with a where clause that contains a constant that is saturated? Can you give an example when you say MDB does not do anything or return a warning?

File: tests/dataconvert-tests.cpp (hunk 3 above)
Comment (tntnatbry) [reply tests/dataconvert-tests.cpp]: No specific reason, I wanted to use the same scale values (0, 10, 38) that I was using for other tests to make it consistent across the test suite :)

File: utils/dataconvert/dataconvert.cpp (hunk 4 above)
Comment (tntnatbry) [reply utils/dataconvert/dataconvert.cpp]: I did not understand the comment. All the case arguments are unparanthesised.

**Comment by tntnatbry** [comment]
I have forced push the PR to include the rebase on latest HEAD, and also added a UT for copyBinaryField.

File: dbcon/joblist/jlf_execplantojoblist.cpp (hunk 1 above)
Comment (drrtuy) [reply dbcon/joblist/jlf_execplantojoblist.cpp]: UPDATE t1 set d1 = 42 WHERE d1 > 99999999999999999999999999999999999999999999999999999999999999

File: dbcon/joblist/tupleconstantstep.cpp [hunk 7]
```diff
@@ -535,7 +535,11 @@ void TupleConstantStep::fillInConstants(const rowgroup::Row& rowIn, rowgroup::Ro
 
         //fRowConst.copyField(rowOut.getData()+2, 0); // hardcoded 2 for rid length
         for (uint32_t i = 1; i < rowOut.getColumnCount(); i++)
-            rowIn.copyField(rowOut, i, i - 1);
+            // WIP MCOL-641 implement copyBinaryField inside copyField
+            if (UNLIKELY(rowIn.getColumnWidth(i - 1) == 16))
```
Comment (drrtuy) [inline dbcon/joblist/tupleconstantstep.cpp]: We can use isWide here. As of now any wide column uses binary ops from RowGroup.

File: tests/dataconvert-tests.cpp (hunk 3 above)
Comment (drrtuy) [reply tests/dataconvert-tests.cpp]: I disagree that we need to remove scale = 37. This allows us to check second case branch in toString.

File: tests/rowgroup-tests.cpp [hunk 8]
```diff
@@ -30,125 +31,138 @@ using int128_t = __int128;
...
+
+        rg.initRow(&r);
+        rg.initRow(&rOutMappingCheck);
+        rowSize = r.getSize();
+        rg.getRow(0, &r);
+
+        int128_t nullValue = 0;
+        int128_t bigValue = 0;
+
+        utils::setWideDecimalNullValue(nullValue);
```
Comment (drrtuy) [inline tests/rowgroup-tests.cpp]: This is not the original  but feel free to add you test as well.

File: utils/dataconvert/dataconvert.cpp (hunk 4 above)
Comment (drrtuy) [reply utils/dataconvert/dataconvert.cpp]: Now, there is no difference but IIRC the previous version has case parenthised values in writeFractionalPart.

File: utils/dataconvert/dataconvert.cpp [hunk 9]
```diff
@@ -1249,115 +1250,144 @@ bool stringToTimestampStruct(const string& data, TimeStamp& timeStamp, const str
...
-
-  // pod[0] is low 8 byte, pod[1] is high
-  uint64_t* high_pod = reinterpret_cast<uint64_t*>(&high);
-  uint64_t* mid_pod = reinterpret_cast<uint64_t*>(&mid);
-  uint64_t* low_pod = reinterpret_cast<uint64_t*>(&low);
-  char* original_p = p;
-  int written = 0;
+    int128_t intPart = *dec;
+    int128_t high = 0, mid = 0, low = 0;
+    uint64_t div = 10000000000000000000ULL;
```
Comment (drrtuy) [inline utils/dataconvert/dataconvert.cpp]: github doesn't like this keyword me neither.
http://www.cplusplus.com/reference/cstdlib/div/

File: utils/dataconvert/dataconvert.cpp [hunk 10]
```diff
@@ -1249,115 +1250,144 @@ bool stringToTimestampStruct(const string& data, TimeStamp& timeStamp, const str
...
+
+    while (scaleDivisor > 1 && fractionalPart / scaleDivisor == 0)
     {
         *p++ = '0';
         written++;
         scaleDivisor /= 10;
     }
-    written += writeIntPart(&fractionalPart, p, buflen, 0);
+
+    p += writeIntPart(&fractionalPart, p, buflen - written, 0);
```
Comment (drrtuy) [inline utils/dataconvert/dataconvert.cpp]: Hm, could you elaborate on what is the reason behind the change from written += writeIntPart to p += writeIntPart ?

File: utils/rowgroup/rowgroup.cpp [hunk 11]
```diff
@@ -593,22 +593,23 @@ string Row::toString() const
...
                 }
+                // WIP MCOL-641
                 case CalpontSystemCatalog::BINARY:
                     std::cout << __FILE__<< ":" <<__LINE__ << " Fix for 16 Bytes ?" << std::endl;
                     break;
-                // WIP
                 case CalpontSystemCatalog::DECIMAL:
                 case CalpontSystemCatalog::UDECIMAL:
-                    if (colWidths[i] == sizeof(int128_t))
+                    if (utils::isWide(colWidths[i]))
```
Comment (drrtuy) [inline utils/rowgroup/rowgroup.cpp]: We can't use isWide here we need to compare with 16.

File: utils/rowgroup/rowgroup.cpp [hunk 12]
```diff
@@ -593,22 +593,23 @@ string Row::toString() const
...
                     {
-                        char *buf = (char*)alloca(precision[i] + 3);
+                        unsigned int buflen = precision[i] + 3;
```
Comment (drrtuy) [inline utils/rowgroup/rowgroup.cpp]: Max length is 42 not 41. Kindly use utils::MAXLENGTH16BYTES here.

File: utils/rowgroup/rowgroup.h (hunk 6 above)
Comment (drrtuy) [reply utils/rowgroup/rowgroup.h]: Nice one.

File: utils/rowgroup/rowgroup.h [hunk 13]
```diff
@@ -1188,6 +1191,12 @@ inline void Row::copyField(Row& out, uint32_t destIndex, uint32_t srcIndex) cons
         out.setIntField(getIntField(srcIndex), destIndex);
 }
 
+// WIP MCOL-641
+inline void Row::copyBinaryField(Row& out, uint32_t destIndex, uint32_t srcIndex) const
+{
+    out.setBinaryField<int128_t>(getBinaryField<int128_t>(srcIndex), 16, destIndex);
```
Comment (drrtuy) [inline utils/rowgroup/rowgroup.h]: Minor thing but I think we don't need explicit parameter here given that getBinaryField returns int128_t*.

File: dbcon/joblist/jlf_execplantojoblist.cpp (hunk 1 above)
Comment (tntnatbry) [reply dbcon/joblist/jlf_execplantojoblist.cpp]: This is good to know. I verified that in strict mode, an out of range value is errored out with error "ERROR 1264 (22003): CAL0002: IDB-2025: Data truncated for column 'a'", while in non-strict mode, the value is saturated with the error converting to a warning.

Comment (drrtuy) [reply dbcon/joblist/jlf_execplantojoblist.cpp]: I can embrace the fact that CS'es behavior differs from MDB's on this matter.

File: utils/dataconvert/dataconvert.cpp (hunk 10 above)
Comment (tntnatbry) [reply utils/dataconvert/dataconvert.cpp]: 3rd argument to writeIntPart was buflen earlier. This was incorrect, since the available buffer should be buflen - written

File: tests/dataconvert-tests.cpp (hunk 3 above)
Comment (tntnatbry) [reply tests/dataconvert-tests.cpp]: Added back the test for scale=37.

File: dbcon/joblist/tupleconstantstep.cpp (hunk 7 above)
Comment (tntnatbry) [reply dbcon/joblist/tupleconstantstep.cpp]: Done.

File: utils/rowgroup/rowgroup.cpp (hunk 11 above)
Comment (tntnatbry) [reply utils/rowgroup/rowgroup.cpp]: Changed it to datatypes::MAXDECIMALWIDTH.

File: utils/rowgroup/rowgroup.cpp (hunk 12 above)
Comment (tntnatbry) [reply utils/rowgroup/rowgroup.cpp]: Done.

File: utils/rowgroup/rowgroup.h (hunk 13 above)
Comment (tntnatbry) [reply utils/rowgroup/rowgroup.h]: Good point, done.
